### PR TITLE
Add MaxPoolSingleOut to ONNXToTOSA

### PR DESCRIPTION
**Pull request by flemairen6**

### CODE REVIEW

File: src/Conversion/ONNXToTOSA/NN/MaxPoolSingleOut.cpp [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+//
+// This file lowers ONNX MaxpoolSingleOut operator to TOSA dialect.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Dialect/Tosa/IR/TosaOps.h"
+#include "mlir/IR/Attributes.h"
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/IR/TypeUtilities.h"
+#include "mlir/IR/Value.h"
```
Comment (maxbartel) [inline src/Conversion/ONNXToTOSA/NN/MaxPoolSingleOut.cpp:19]: Could you clean this a bit up? I think the language server adds most of those

File: src/Conversion/ONNXToTOSA/NN/MaxPoolSingleOut.cpp [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+  LogicalResult matchAndRewrite(ONNXMaxPoolSingleOutOp op, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+
+    Location loc = op->getLoc();
+
+    Value Input = adaptor.X();
+    // - The attribute Ceil_mode is ignored because its effect is already impacting the return
+    // type of the operator
+    // - The attribute auto_pad is ignored because it is marked as deprecated for ONNX
+    // - The attributes storage_order and dilations is also unsupported
```
Comment (maxbartel) [inline src/Conversion/ONNXToTOSA/NN/MaxPoolSingleOut.cpp:43]: I think it is not good to ignore them. If some model has them set, the debugging will be very hard. Better check for the default values and use `op.emitError` or `rewriter.notifyMatchFailure` if they differ.

File: src/Conversion/ONNXToTOSA/NN/MaxPoolSingleOut.cpp [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    mlir::StringAttr autoPad = adaptor.auto_padAttr();
+    
+    mlir::ArrayAttr dilations = adaptor.dilationsAttr();
+    mlir::ArrayAttr kernelShape = adaptor.kernel_shapeAttr();
+    mlir::ArrayAttr pads = adaptor.padsAttr();
+    mlir::ArrayAttr strides = adaptor.stridesAttr();
+
+    auto InputType = Input.getType().cast<RankedTensorType>();
```
Comment (maxbartel) [inline src/Conversion/ONNXToTOSA/NN/MaxPoolSingleOut.cpp:51]: The input type can also be a memref. This line should catch that, but maybe add a comment saying memrefs are not supported

File: src/Conversion/ONNXToTOSA/NN/MaxPoolSingleOut.cpp [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    auto resultType = op.getResult().getType();
+    auto resultShape = resultType.cast<RankedTensorType>().getShape();
+    // Construct the transposed type for the new MaxPool OP
+    Type newResultType = RankedTensorType::get(
+      {resultShape[0], resultShape[2], resultShape[3], resultShape[1]}, InputType.getElementType());
+
+    if (autoPad != "NOTSET") {
```
Comment (maxbartel) [inline src/Conversion/ONNXToTOSA/NN/MaxPoolSingleOut.cpp:58]: MLIR has a different mechanism for these type of print statements. See [MLIR doc](https://mlir.llvm.org/getting_started/Debugging/#:~:text=Miscellaneous%20tips%C2%A0,on%2Ddiagnostic.). Aside, I think the conversion should better fail if those are set.

File: src/Conversion/ONNXToTOSA/NN/MaxPoolSingleOut.cpp [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    }
+    if (dilations) {
+      llvm::errs() << "dilations attribute is unsupported by TOSA and its value will be ignored.";
+    }
+    
+    // ONNX Mlir uses NCHW as an input while TOSA expects NHWC. Insert a transpose
+    // to change the format
+    Value targetTensor = tosa::getConstTensor<int32_t>(rewriter, op, {0, 2, 3, 1}, {4}).value();
+    Type outputType = RankedTensorType::get({-1, -1, -1, -1}, InputType.getElementType());
+    Input = tosa::CreateOpAndInfer<tosa::TransposeOp>(rewriter, loc, outputType, Input, targetTensor).getResult();
```
Comment (maxbartel) [inline src/Conversion/ONNXToTOSA/NN/MaxPoolSingleOut.cpp:69]: The conv merge introduced small helper functions for this, but it is up to you if you want to use them. They don't do anything special.

File: src/Conversion/ONNXToTOSA/NN/MaxPoolSingleOut.cpp [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+
+    // Pads and Strides are optionals for ONNX but mandatory for TOSA
+    if (!pads) {
+      pads = rewriter.getI64ArrayAttr({0, 0, 0, 0});
+    }
+    else {
+      llvm::SmallVector<int64_t, 4> transposedPads = extractFromI64ArrayAttr(pads);
+      pads = rewriter.getI64ArrayAttr(
+          {transposedPads[0], transposedPads[2], transposedPads[3], transposedPads[1]});
```
Comment (maxbartel) [inline src/Conversion/ONNXToTOSA/NN/MaxPoolSingleOut.cpp:78]: I think 3 and 1 should be switched.

File: test/mlir/conversion/onnx_to_tosa/NN/MaxPoolSingleOut.mlir [hunk 7]
```diff
@@ -0,0 +1,103 @@
+// RUN: onnx-mlir-opt --shape-inference --convert-onnx-to-tosa %s -split-input-file | FileCheck %s
+
+/// Test the default behavior of Max Pool with no padding
+func.func @test_default_maxpoolsingleout(%arg0 : tensor<5x5x32x32xf32>) -> tensor<5x5x30x30xf32> {
+  %0 = "onnx.MaxPoolSingleOut"(%arg0) {kernel_shape = [3,3]} : (tensor<5x5x32x32xf32>) -> tensor<5x5x30x30xf32>
```
Comment (maxbartel) [inline test/mlir/conversion/onnx_to_tosa/NN/MaxPoolSingleOut.mlir:5]: Did you run the pass with the `--debug` flag and checked if ONNX or TOSA emitted shape warnings? Just out of curiosity how you got all the shapes. That was the main reason I adapted your dynamic shape + infer combination, but you set all shapes explicitly

File: test/mlir/conversion/onnx_to_tosa/NN/MaxPoolSingleOut.mlir [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+// CHECK-DAG:     %[[MPOOL_RES:.*]] = "tosa.max_pool2d"(%[[TRANS_ARG]]) {kernel = [3, 3], pad = [1, 1, 1, 1], stride = [1, 1]} : (tensor<5x32x32x5xf32>) -> tensor<5x32x32x5xf32>
+// CHECK-DAG:     "tosa.const"() {value = dense<[0, 3, 1, 2]> : tensor<4xi32>} : () -> tensor<4xi32>
+// CHECK-DAG:     %[[TRANS_MPOOL_RES:.*]] = "tosa.transpose"(%[[MPOOL_RES]], %3) : (tensor<5x32x32x5xf32>, tensor<4xi32>) -> tensor<5x5x32x32xf32>
+// CHECK-DAG:     return %[[TRANS_MPOOL_RES]] : tensor<5x5x32x32xf32>
+
+// -----
+
+/// Test the behavior of Max Pool with non uniform padding
+func.func @test_default_maxpoolsingleout_pad_nonunif(%arg0 : tensor<5x5x32x32xf32>) -> tensor<5x5x31x31xf32> {
+  %0 = "onnx.MaxPoolSingleOut"(%arg0) {ceil_mode = 0 : si64, kernel_shape = [5,3], pads = [2, 1, 1, 0] } : (tensor<5x5x32x32xf32>) -> tensor<5x5x31x31xf32>
```
Comment (maxbartel) [inline test/mlir/conversion/onnx_to_tosa/NN/MaxPoolSingleOut.mlir:35]: Could you change the pad vector to something like [0, 1, 2, 3]? Then we are sure that the reordering is correct

File: src/Conversion/ONNXToTOSA/NN/MaxPoolSingleOut.cpp (hunk 4 above)
Comment (flemairen6) [reply src/Conversion/ONNXToTOSA/NN/MaxPoolSingleOut.cpp:58]: Good catch, I didn't know about it

File: test/mlir/conversion/onnx_to_tosa/NN/MaxPoolSingleOut.mlir (hunk 7 above)
Comment (flemairen6) [reply test/mlir/conversion/onnx_to_tosa/NN/MaxPoolSingleOut.mlir:5]: If you are talking about the ONNX mlir code, I got it from other tests mostly. 
Though, I ran my tests with `--debug` and saw some warning. I'm investigating them right now, but I also saw that fully relying on shape inference failed in TOSA, hence why I'm calling `rewriter.create` to create the MaxPool and not `CreateOpAndInfer`

Comment (flemairen6) [reply test/mlir/conversion/onnx_to_tosa/NN/MaxPoolSingleOut.mlir:5]: Update on that, it was a poorly written test that was generating warning. So, no warning, now that it's fixed

**Review by maxbartel — APPROVED** [review]

**Pull request merged**